### DISCUSSION TITLE
How can I run my bot using webhooks instead of polling

### DISCUSSION THREAD

**Discussion by SimaTheCoder** [telegraf/telegraf#1672] [Q&A] [answered]
Hey

I'm using node.js,mysql,javascript. How can I add webhooks to my telegram bot? Please provide an example or a  help link.
I've tried following the examples but they don't seem to work.

I would also like to know. Is there really a difference in performance/speed when it comes to webhooks and polling?
Is there anyone using polling and able to manage multiple users requests?
Which port is recommended?

Your assistance will be highly appreciated!

**Comment by MKRhere** [comment, marked as answer]
Example: https://github.com/telegraf/telegraf/blob/v4/docs/examples/webhook-bot.js

> Is there really a difference in performance/speed

This should only really be a concern if you're running a bot with a high amount of traffic per second. Most bots will not need to worry about setting up webhooks.

> Which port is recommended?

Very similar to running a webserver (because this is a webserver), you can listen on any localhost port, and reverse-proxy it with a server like nginx. If you really wish to your bot _directly_ on the internet, then you must use port 443 and pass TLS options like so: `bot.launch({ webhook: { ..., port: 443, tlsOptions: { ... } } })`. tlsOptions are the same as node's https module. This is not recommended, however.

> able to manage multiple users requests

A long-polling bot (default) will be able to serve several thousand users a day with no problem. You might want to start using webhooks if you notice significantly bad performance due to a large number of users per second.

**Reply by alexey-sh** [reply]
Your example link is broken. Would you mind to update it?

**Reply by MKRhere** [reply]
https://github.com/feathers-studio/telegraf-docs/tree/6be310123a1d39c052c45823e640ebaab22ea9af/examples/webhook

**Comment by SimaTheCoder** [comment]
Thank you for your response and assistance. I really appreciate it.

So is this the only code required in order to successfully add a webhook to a bot? 
```
bot.launch({
  webhook:{
    host:100.100.100.42,
    domain: "https://api.telegram.org/bot<bot_token>/setWebhook?url=https://bot.example.co.za",
  	port:443,
  }
})
```

Whenever I use port 443 I get the following error:

```
Error: listen EACCES: permission denied 100.100.100.42:443
 code: 'EACCES',
  errno: -4092,
  syscall: 'listen',
  address: 'ip_address',
  port: 443
```

Whenever I change the port to 3000 or 8080, I don't get this error but the bot doesn't respond to my messages.

I'm not using reverse proxy so that could be the reason. Could you please tell me how I can add reverse-proxy.

The bot is meant to be used by 200-300 users and It's currently being tested.
I'm running the bot using polling + adding it as a windows service.
One of the users has been complaining though about how slow it is to respond.
Could it be the network connection or the mysql connection or just the polling?

**Reply by MKRhere** [reply]
`domain` should just be `bot.example.co.za`. Port 443 is restricted to admin, so you'd have to run with root/admin privileges (again, highly, highly not recommended, you should just use a proxy server like nginx).

> 200-300 users

You almost definitely can use long-polling. Slow response time is not from polling, but something else in your pipeline as you suspect.

> how I can add reverse-proxy

This is a more involved answer... I would suggest searching for "how to run a node.js server with nginx", or on those lines.

> as a windows service

I actually have no knowledge about running servers on Windows. I don't know if this helps, but there is [nginx for Windows](http://nginx.org/en/docs/windows.html) and one of my reverse-proxy guides (untested on Windows):
* [Using DNS (better)](https://gist.github.com/MKRhere/e9a427aa3febd9215098db17f240467f)
* [Using stateless configuration (easier)](https://gist.github.com/MKRhere/09fb6d1e6a338b180c28c8a93e7a60bc)

You might also have better luck just searching for this.

TL;DR I don't recommend webhooks if you're running from a home computer, etc and don't really need to serve a high volume of users. Benchmark different parts of your bot to see what's actually slowing down your bot.

**Reply by SimaTheCoder** [reply]
Okay noted thank you so much. This is so helpful and informative. I highly appreciate your assistance.
